airflow/models.py

```diff
@@ -3626,6 +3629,52 @@ def delete(cls, xcoms, session=None):
         session.commit()
 
 
+class DagStat(Base):
+    __tablename__ = "dag_stats"
+
+    dag_id = Column(String(ID_LEN), primary_key=True)
+    state = Column(String(50), primary_key=True)
+    count = Column(Integer, default=0)
+    dirty = Column(Boolean, default=False)
+
+    def __init__(self, dag_id, state, count, dirty=False):
+        self.dag_id = dag_id
+        self.state = state
+        self.count = count
+        self.dirty = dirty
+
+    @staticmethod
+    @provide_session
+    def set_dirty(dag_id, session=None):
+        for dag in session.query(DagStat).filter(DagStat.dag_id == dag_id):
+            dag.dirty = True
+        session.commit()
+
+    @staticmethod
+    @provide_session
+    def clean_dirty(session=None):
+        dirty_dags = {dag.dag_id for dag in session.query(DagStat).filter_by(dirty=True)}
+        session.query(DagStat).filter_by(dirty=True).delete()
+        qry = None
+        if not dirty_dags:
+            qry = (
+                session.query(DagRun.dag_id, DagRun.state, func.count('*'))
+                .filter(or_(~DagRun.dag_id.in_(session.query(DagStat.dag_id))))
```

Comment:
`DagRun.dag_id` may not be efficient. Currently, there is a btree index on the combination of dag_id and  execution_date in the dag_run table. For companies with 1M or more dag_runs, this will require scanning the index for much less than 1M entries. Maybe we can get the dag_ids from the dagbag.